src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -350,4 +370,205 @@ internal static bool TryParseSeverity(string value, out ReportDiagnostic severit
             return false;
         }
     }
+
+    /// <summary>
+    /// Represents a global key that was unset due to conflicts
+    /// </summary>
+    public readonly struct AnalyzerUnsetKey
+    {
+        /// <summary>
+        /// The name of the key that was unset
+        /// </summary>
+        public string KeyName { get; }
+
+        /// <summary>
+        /// The section this key came from 
+        /// </summary>
+        /// <remarks>
+        /// Will be 'Global Section' when <see cref="IsGlobalSection"/> is <c>true</c>.
+        /// </remarks>
+        public string SectionName { get; }
+
+        /// <summary>
+        /// <c>true</c> if the section this key came from was the global section
+        /// </summary>
+        public bool IsGlobalSection { get; }
+
+        /// <summary>
+        /// The list of paths of the global configs that each set this key
+        /// </summary>
+        public ImmutableArray<string> ConfigPaths { get; }
+
+        internal AnalyzerUnsetKey(string keyName, string sectionName, bool isGlobalSection, ImmutableArray<string> paths)
+        {
+            KeyName = keyName;
+            SectionName = sectionName;
+            IsGlobalSection = isGlobalSection;
+            ConfigPaths = paths;
+        }
+    }
+
+    /// <summary>
+    /// Builds a global analyzer config from a series of configs
+    /// </summary>
+    internal class GlobalAnalyzerConfigBuilder
```

Comment:
Why do we need this type? Why can't we use an AnalyzerConfigSet, which should know what to do with global configs?

Reply:
I think this comment still stands. I think we should have a uniform way of viewing a group of analyzer configs, and that representation is currently the AnalyzerConfigSet. If we want to pre-process global configs inside the analyzer config set that's fine, but then I think this should be a nested private type.

Reply:
That's what we're doing. The type was just exposed as internal so that we can test it more easily. I can move it into the configset though.

---
In reply to: [421830516](https://github.com/dotnet/roslyn/pull/43889#discussion_r421830516) [](ancestors = 421830516)